src/ImageSharp/Common/Helpers/SimdUtils.Shuffle.cs

```diff
@@ -0,0 +1,150 @@
+// Copyright (c) Six Labors.
+// Licensed under the Apache License, Version 2.0.
+
+using System;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+namespace SixLabors.ImageSharp
+{
+    internal static partial class SimdUtils
+    {
+        /// <summary>
+        /// Shuffle single-precision (32-bit) floating-point elements in <paramref name="source"/>
+        /// using the control and store the results in <paramref name="dest"/>.
+        /// </summary>
+        /// <param name="source">The source span of floats.</param>
+        /// <param name="dest">The destination span of floats.</param>
+        /// <param name="control">The byte control.</param>
+        [MethodImpl(InliningOptions.ShortMethod)]
+        public static void Shuffle4Channel(
+            ReadOnlySpan<float> source,
+            Span<float> dest,
+            byte control)
+        {
+            VerifyShuffleSpanInput(source, dest);
+
+            // TODO: There doesn't seem to be any APIs for
+            // System.Numerics that allow shuffling.
+#if SUPPORTS_RUNTIME_INTRINSICS
+            HwIntrinsics.Shuffle4ChannelReduce(ref source, ref dest, control);
+#endif
+
+            // Deal with the remainder:
+            if (source.Length > 0)
+            {
+                ShuffleRemainder4Channel(source, dest, control);
+            }
+        }
+
+        /// <summary>
+        /// Shuffle 8-bit integers in a within 128-bit lanes in <paramref name="source"/>
+        /// using the control and store the results in <paramref name="dest"/>.
+        /// </summary>
+        /// <param name="source">The source span of bytes.</param>
+        /// <param name="dest">The destination span of bytes.</param>
+        /// <param name="control">The byte control.</param>
+        [MethodImpl(InliningOptions.ShortMethod)]
+        public static void Shuffle4Channel(
+            ReadOnlySpan<byte> source,
+            Span<byte> dest,
+            byte control)
+        {
+            VerifyShuffleSpanInput(source, dest);
+
+            // TODO: There doesn't seem to be any APIs for
+            // System.Numerics that allow shuffling.
+#if SUPPORTS_RUNTIME_INTRINSICS
+            HwIntrinsics.Shuffle4ChannelReduce(ref source, ref dest, control);
+#endif
+
+            // Deal with the remainder:
+            if (source.Length > 0)
+            {
+                ShuffleRemainder4Channel(source, dest, control);
```

Comment:
The non-SIMD shuffling has optimized implementations in `PixelConverter`, although I'm unsure what runtimes support those optimizations.

Reply:
Yeah, I saw that but there's a limited set of shuffles (Though we could maybe T4 template it?). With this API we could potentially expose the ability to use any shuffle combination. for example `XXXX`.

Reply:
Really depends on where do we want to get in the end, and how the overall pipelines would look like. (See my main comment.)

Reply:
If we want to avoid regression on non-intel and/or non 3.1+, I think we have 2 options now:

1. Handle the cases right in the `PixelOperations` method:

```C#
            public override void FromRgba32(Configuration configuration, ReadOnlySpan<Rgba32> sourcePixels, Span<Argb32> destinationPixels)
            {
                Guard.NotNull(configuration, nameof(configuration));
                Guard.DestinationShouldNotBeTooShort(sourcePixels, destinationPixels, nameof(destinationPixels));

#if SUPPORTS_RUNTIME_INTRINSICS
                if (Avx.IsSupported || Sse.IsSupported)
                {
                    var source = MemoryMarshal.Cast<Rgba32, byte>(sourcePixels);
                    var dest = MemoryMarshal.Cast<Argb32, byte>(destinationPixels);

                    SimdUtils.Shuffle4Channel(source, dest, SimdUtils.Shuffle.WXYZ);
                }
                else
                
#else
                {
                    ref uint sourceRef = ref Unsafe.As<Rgba32, uint>(ref MemoryMarshal.GetReference(sourcePixels));
                    ref uint destRef = ref Unsafe.As<Argb32, uint>(ref MemoryMarshal.GetReference(destinationPixels));

                    for (int i = 0; i < sourcePixels.Length; i++)
                    {
                        uint sp = Unsafe.Add(ref sourceRef, i);
                        Unsafe.Add(ref destRef, i) = PixelConverter.FromRgba32.ToArgb32(sp);
                    }
                }
#endif
            }
```

2. Add a `switch` statement to `ShuffleRemainder4Channel` to handle well-known cases for `control` with the code in `PixelConverter`.

What is the best depends on the plans to handle the 4->3 and the 3->4 cases, so we have consistency in the resulting code. See https://github.com/SixLabors/ImageSharp/pull/1404#issuecomment-717258370

Reply:
2 is probably better

Reply:
Actually there is one other option:
3. T4 multiple variants of Shuffle4Channel, so we can avoid the switch-branch, but it quite a lot of work, unsure if the perf gains will justify it.

Reply:
There you go.... I haven't wired them up yet but I've optimized our existing shuffles so there's no slowdown on older NET Core. 

About 4x faster than they were previously.

|          Method |             Job |                              EnvironmentVariables | Count |      Mean |    Error |   StdDev | Ratio | RatioSD | Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------------- |---------------- |-------------------------------------------------- |------ |----------:|---------:|---------:|------:|--------:|------:|------:|------:|----------:|
| Shuffle4Channel |             AVX |                                             Empty |   128 |  21.56 ns | 0.454 ns | 0.651 ns |  1.00 |    0.00 |     - |     - |     - |         - |
| Shuffle4Channel | No HwIntrinsics | COMPlus_EnableHWIntrinsic=0,COMPlus_FeatureSIMD=0 |   128 |  19.50 ns | 0.252 ns | 0.224 ns |  0.89 |    0.03 |     - |     - |     - |         - |
| Shuffle4Channel |             SSE |                               COMPlus_EnableAVX=0 |   128 |  17.66 ns | 0.125 ns | 0.111 ns |  0.81 |    0.03 |     - |     - |     - |         - |
|                 |                 |                                                   |       |           |          |          |       |         |       |       |       |           |
| Shuffle4Channel |             AVX |                                             Empty |   256 |  24.49 ns | 0.284 ns | 0.252 ns |  1.00 |    0.00 |     - |     - |     - |         - |
| Shuffle4Channel | No HwIntrinsics | COMPlus_EnableHWIntrinsic=0,COMPlus_FeatureSIMD=0 |   256 |  38.02 ns | 0.783 ns | 1.219 ns |  1.51 |    0.04 |     - |     - |     - |         - |
| Shuffle4Channel |             SSE |                               COMPlus_EnableAVX=0 |   256 |  29.11 ns | 0.317 ns | 0.281 ns |  1.19 |    0.02 |     - |     - |     - |         - |
|                 |                 |                                                   |       |           |          |          |       |         |       |       |       |           |
| Shuffle4Channel |             AVX |                                             Empty |   512 |  29.75 ns | 0.233 ns | 0.218 ns |  1.00 |    0.00 |     - |     - |     - |         - |
| Shuffle4Channel | No HwIntrinsics | COMPlus_EnableHWIntrinsic=0,COMPlus_FeatureSIMD=0 |   512 |  79.31 ns | 1.252 ns | 1.171 ns |  2.67 |    0.05 |     - |     - |     - |         - |
| Shuffle4Channel |             SSE |                               COMPlus_EnableAVX=0 |   512 |  29.84 ns | 0.603 ns | 0.592 ns |  1.00 |    0.02 |     - |     - |     - |         - |
|                 |                 |                                                   |       |           |          |          |       |         |       |       |       |           |
| Shuffle4Channel |             AVX |                                             Empty |  1024 |  38.62 ns | 0.244 ns | 0.228 ns |  1.00 |    0.00 |     - |     - |     - |         - |
| Shuffle4Channel | No HwIntrinsics | COMPlus_EnableHWIntrinsic=0,COMPlus_FeatureSIMD=0 |  1024 | 150.61 ns | 0.293 ns | 0.245 ns |  3.90 |    0.02 |     - |     - |     - |         - |
| Shuffle4Channel |             SSE |                               COMPlus_EnableAVX=0 |  1024 |  46.57 ns | 0.606 ns | 0.567 ns |  1.21 |    0.02 |     - |     - |     - |         - |
|                 |                 |                                                   |       |           |          |          |       |         |       |       |       |           |
| Shuffle4Channel |             AVX |                                             Empty |  2048 |  59.99 ns | 0.352 ns | 0.312 ns |  1.00 |    0.00 |     - |     - |     - |         - |
| Shuffle4Channel | No HwIntrinsics | COMPlus_EnableHWIntrinsic=0,COMPlus_FeatureSIMD=0 |  2048 | 283.23 ns | 0.869 ns | 0.770 ns |  4.72 |    0.02 |     - |     - |     - |         - |
| Shuffle4Channel |             SSE |                               COMPlus_EnableAVX=0 |  2048 |  85.03 ns | 1.708 ns | 1.899 ns |  1.42 |    0.04 |     - |     - |     - |         - |